engines/registry/app/controllers/registry/registry_controller.rb

```diff
@@ -0,0 +1,99 @@
+module Registry
+  class RegistryController < ::ApplicationController
+    REGISTRY_SERVICE = 'SUSE Linux OCI Registry'.freeze
+    REGISTRY_API_VERSION = 'registry/2.0'.freeze
+
+    before_action :set_requested_scopes, except: [ :catalog ]
+    before_action :basic_auth, except: [ :catalog ]
+    before_action :catalog_token_auth, only: [ :catalog ]
+
+    rescue_from Exception, with: :handle_exceptions
+
+    # AuthZ handler
+    # AuthZ will validate which of the requested scope policies are fulfilled
+    # with the current login access and prepare the token to be sent back to the client
+    def authorize
+      token = Registry::AccessToken.new(@client&.account, params['service'], @requested_scopes.map { |s| s.granted(client: @client) }).token
+      render json: { token: token }, status: :ok
+    end
+
+    # Catalog handler
+    # Returns a Distribution Registry HTTP API V2 - compatible repository catalog as defined in
+    # https://distribution.github.io/distribution/spec/api/#listing-repositories
+    def catalog
+      repos = @client.matching_policies.repositories(account_numbers: @client.account_numbers)
+      logger.debug("Returning #{repos.size} repos for client #{@client}")
+
+      response.set_header('Docker-Distribution-Api-Version', REGISTRY_API_VERSION)
+      render json: { repositories: repos }, status: :ok
+    end
+
+    private
+
+    # Support multiple scopes. Podman & Docker handle this differently
+    #   - Podman sends multiple querystrings with the same name.
+    #   - Docker sends space-separated values in only one query string.
+    # This should normalize everything
+    def set_requested_scopes
+      raw_scopes = CGI.parse(request.env['QUERY_STRING']).fetch('scope', [])
+      .join(' ')
+      .split
+      .map(&:presence)
+      .compact
+      .map(&:downcase)
+
+      @requested_scopes = []
+      unless raw_scopes.empty?
+        @requested_scopes = raw_scopes.map do |scope| #.filter_map do |scope|
+          Registry::AccessScope.parse(scope)
+        end
+      end
+      logger.info("Requested scopes: #{@requested_scopes.map(&:to_s)}")
+    end
+
+    # AuthN handler
+    # https://docs.docker.com/registry/spec/auth/jwt/#getting-a-bearer-token
+    def basic_auth
+      # skip authentication if this is not a login request
+      return unless request.authorization
+
+      authenticate_or_request_with_http_basic('SUSE Registry Authentication') do |login, password|
+        begin
+          @client = Registry::AuthenticatedClient.new(login, password)
+        rescue StandardError
+          logger.info _('Could not find system with login \"%{login}\" and password \"%{password}\"') %
+            { login: login, password: password }
```

Comment:
his message does not help the user, it only tells the user that they are SOL. We need to tell the user what to do. In the special case we are currently implementing, i.e. for the public cloud update infrastructure only this should read.
"Registration expired please run cloudregistryauth command".  For air-gaped environments auth probably doesn't matter and our specialized message may be sufficient, but that requires more thought later. For now let's make sure we tell the user what they need to do.

Reply:
I'm don't think podman will show custom messages to the user at all. If I recall correctly it's showing it's own messages depending on the reponse code.

Reply:
It does, it would show 
```bash
> podman search registry-ec2.susecloud.net/
ERRO[0001] error getting search results from v2 endpoint "registry-ec2.susecloud.net": requested access to the resource is denied 
Error: 1 error occurred:
	* couldn't search registry "registry-ec2.susecloud.net": requested access to the resource is denied
```
that log information is for internal use, not for the user, happy to change the log message though

Reply:
@rjschwei what would `cloudregistryauth` command be ? 
We get the credentials when registering the instance on the client, 
I understand the command would be `registercloudguest` but, do we want the user
to re register the system to use (re-auth on) the registry ?